### PR TITLE
Add admin enforcement to protected branches API (#606)

### PR DESCRIPTION
**Pull request by amirilovic**
Helps #606.

### CODE REVIEW

File: github/repos.go [hunk 1]
```diff
@@ -511,13 +511,15 @@ type Branch struct {
 type Protection struct {
 	RequiredStatusChecks       *RequiredStatusChecks       `json:"required_status_checks"`
 	RequiredPullRequestReviews *RequiredPullRequestReviews `json:"required_pull_request_reviews"`
+	EnforceAdmins              *AdminEnforcement           `json:"enforce_admins"`
 	Restrictions               *BranchRestrictions         `json:"restrictions"`
```
Comment (shurcooL) [inline github/repos.go]: The names of other structs here take pretty closely after the field name itself.

Have you considered naming this struct something like `EnforceAdmins`? If so, what made you pick `AdminEnforcement` instead?

It seems like the latter sounds better, but I'm a little unsure about the deviation from the pattern. Do you have any comments on this?

I'm on the fence about this, that's why I want to ask for your thoughts.

File: github/repos.go [hunk 2]
```diff
@@ -538,6 +540,12 @@ type RequiredPullRequestReviews struct {
 	IncludeAdmins bool `json:"include_admins"`
 }
 
+// AdminEnforcement represents the configuration to enforce required status checks for repository administrators.
+type AdminEnforcement struct {
+	URL     *string `json:"url,omitempty"`
+	Enabled bool    `json:"enabled"`
```
Comment (shurcooL) [inline github/repos.go]: What's the reason `URL` is a pointer `*string` and has `,omitempty`, unlike `Enabled`?

I think I get the reason, but it seems a bit arbitrary. I don't have any suggestions for improvement, so this is okay with me.

File: github/repos.go [hunk 3]
```diff
@@ -538,6 +540,12 @@ type RequiredPullRequestReviews struct {
 	IncludeAdmins bool `json:"include_admins"`
```
Comment (shurcooL) [inline github/repos.go]: In order to fully resolve #606, I think you should mark `IncludeAdmins` as deprecated, or remove it.

Same for `IncludeAdmins` inside `RequiredStatusChecks` struct.

If you deprecate it, then change it to "Helps #606", so it won't be closed. We want to use that issue to track the entire change for https://developer.github.com/changes/2017-03-22-protected-branches-admin-improvements/, which includes removing the old way of setting admin.

**Comment by shurcooL** [comment]
Whoops, I self-requested a review as a way of not forgetting to deal with this PR, but I forgot anyway. Thanks for reminding.

File: github/repos.go (hunk 1 above)
Comment (amirilovic) [inline github/repos.go]: I choose AdminEnforcement because of: https://developer.github.com/v3/repos/branches/#get-admin-enforcement-of-protected-branch
The GET request states: Get **admin enforcement** of protected branch, which makes sense because it is actually a nested resource, not just a flag.

File: github/repos.go (hunk 3 above)
Comment (amirilovic) [inline github/repos.go]: @shurcooL I'm not sure about deprecation, seems like an overhead, maybe it's better just to remove old fields.

Comment (shurcooL) [inline github/repos.go]: Feel free to go for it and make that a part of this PR.

File: github/repos.go (hunk 1 above)
Comment (shurcooL) [inline github/repos.go]: Sounds good, that's convincing rationale, thanks for providing it.

File: github/repos.go (hunk 2 above)
Comment (navarrano) [inline github/repos.go]: @amirilovic, if you don't mind, I'd love to know the response for shurcooL question, because I'd like to understand your approach here.

Comment (shurcooL) [inline github/repos.go]: @navarrano See https://github.com/google/go-github/issues/537 and the issues it mentions for background information on this.

The TLDR is that this is a style choice, both ways work. Using pointers allows users to know whether the value was actually present in the GitHub response or not, but they have to deal with a pointer. Using values means you can't know if it's a zero value or it was present in GH response, but no pointers to deal with.

My understanding is that `Enabled` is supposed to be a field that's always included in the response, so it's kinda okay not making it a pointer. But URL is expected to be more optional, and it might be worth knowing if it's included or not.

It also has to do with consistency with other similar structs for this API.

But it could've gone in another way.

Comment (amirilovic) [inline github/repos.go]: @shurcooL @navarrano sorry for late response, @shurcooL that was exactly my reasoning, don't know if you had similar cases, but it seemed logical to me.

File: github/repos.go (hunk 3 above)
Comment (amirilovic) [inline github/repos.go]: @shurcooL struct `RequiredPullRequestReviews` should be removed completely?

Comment (shurcooL) [inline github/repos.go]: As I understand, we want to remove `IncludeAdmins` fields. I see that `RequiredPullRequestReviews` contains _only_ that field, and no others, so I understand where your question is coming from.

I would suspect that yes, we want to remove `RequiredPullRequestReviews`, since it has no other fields left, but the GitHub API docs say [some strange/conflicting things](https://developer.github.com/v3/repos/branches/#parameters-1) right now:

> GitHub recommends passing `required_pull_request_reviews` and `enforce_admins`. Passing these fields is not currently required, but may be required soon.

I'm not sure if that's out of date or a typo...

Comment (amirilovic) [inline github/repos.go]: I'll contact support and try out how API behaves currently if no value is set for `RequiredPullRequestReviews`

Comment (amirilovic) [inline github/repos.go]: @shurcooL No reply from support yet but here are test results:
1. When required_pull_request_reviews is not in the request, response status is 200 OK - 
existing required_pull_request_reviews is not affected
1. When required_pull_request_reviews is null, response is 200 OK - existing required_pull_request_reviews is removed from the resource
1. When required_pull_request_reviews is {}, response is 422 Unprocessable Entity, with following error in body:
```
{
  "message": "Invalid request.\n\nNo subschema in \"anyOf\" matched.\n\"include_admins\" wasn't supplied.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', {} is not a null.",
  "documentation_url": "https://developer.github.com/v3/repos/branches/#update-branch-protection"
}
```
So, it's safe to remove required_pull_request_reviews from the request completely. 

For required_status_checks:
1. When required_status_checks is sent with include_admins, , response is 200 OK - existing required_status_checks is updated with new values
1. When required_status_checks is sent without include_admins, response is 422 Unprocessable Entity, with following error in body:
```
{
  "message": "Invalid request.\n\nNo subschema in \"anyOf\" matched.\n\"include_admins\" wasn't supplied.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', {\"url\"=>\"https://api.github.com/repos/amirilovic/test-repo/branches/master/protection/required_status_checks\", \"strict\"=>true, \"contexts\"=>[], \"contexts_url\"=>\"https://api.github.com/repos/amirilovic/test-repo/branches/master/protection/required_status_checks/contexts\"} is not a null.",
  "documentation_url": "https://developer.github.com/v3/repos/branches/#update-branch-protection"
}
```

So, required_status_checks has to include include_admins field.

File: github/repos.go [hunk 4]
```diff
@@ -509,22 +509,21 @@ type Branch struct {
...
-	Restrictions               *BranchRestrictionsRequest  `json:"restrictions"`
+	RequiredStatusChecks *RequiredStatusChecks      `json:"required_status_checks"`
+	EnforceAdmins        bool                       `json:"enforce_admins"`
+	Restrictions         *BranchRestrictionsRequest `json:"restrictions"`
 }
 
 // RequiredStatusChecks represents the protection status of a individual branch.
 type RequiredStatusChecks struct {
+	// Deprecated: Use EnforceAdmins instead.
 	// Enforce required status checks for repository administrators. (Required.)
```
Comment (shurcooL) [inline github/repos.go]: You can't start the comment with "Deprecated:" first, because that breaks the Go style requirement of having documentation begin with identifier name. Put it second:

```
// Enforce required status checks for repository administrators. (Required.)
// Deprecated: Use EnforceAdmins instead.
```

Reference: https://github.com/golang/go/issues/10909#issuecomment-136492606.

**Comment by amirilovic** [comment]
@shurcooL I've fixed the comment issue. I think there is no need to wait for response from support, I've tested current behavior of the API. If something comes up from GH support, I'll report the issue and fix it.

File: github/repos.go (hunk 3 above)
Comment (shurcooL) [inline github/repos.go]: The `RequiredStatusChecks.IncludeAdmins` field is now documented both as required and deprecated.

It's pretty puzzling to me how to use the new API. There are two `bool`s, do they need to be set to the same value if `true`, or does `EnforceAdmins` take higher precedence over `RequiredStatusChecks.IncludeAdmins`? Or can either one be set to `true`?

**Comment by shurcooL** [comment]
Okay. I tried to review the PR and came up with a question at https://github.com/google/go-github/pull/610#discussion_r110760785.

**Comment by shurcooL** [comment]
Also, see #617, which might be relevant here.

It seems to add `dismissal_restrictions` to `required_pull_request_reviews`, so perhaps we don't want to remove `required_pull_request_reviews` anymore.

File: github/repos.go (hunk 3 above)
Comment (amirilovic) [inline github/repos.go]: @shurcooL I've checked how it works. When setting either `RequiredStatusChecks.IncludeAdmins`,`RequiredPullRequestReviews.IncludeAdmins` or `EnforceAdmins` will set all other properties to the same value - changing one changes all the rest. `RequiredPullRequestReviews.IncludeAdmins` and `RequiredPullRequestReviews.IncludeAdmins` are required fields in terms of request structure correctness (request fails with `422 Unprocessable Entity`, if they are not included - explained in previous comment), but logically they shouldn't be used since `EnforceAdmins` overrides them both. So, they are required, but deprecated by `EnforceAdmins`.

**Comment by amirilovic** [comment]
@shurcooL I've rolled back changes removing `RequiredPullRequestReviews`, also added deprecation comment for `RequiredPullRequestReviews.IncludeAdmins`. I've responded to your question about relationship between `RequiredPullRequestReviews.IncludeAdmins`, `RequiredStatusChecks.IncludeAdmins` and `EnforceAdmins`.

**Comment by amirilovic** [comment]
@shurcooL Anything else that needs to be done here?

**Comment by shurcooL** [comment]
Thanks for making progress on this. I'll try to review this soon.

File: github/repos.go (hunk 3 above)
Comment (shurcooL) [inline github/repos.go]: I still think this is confusing Go API for go-github users. But, this falls on GitHub, and it's temporary. According to https://developer.github.com/changes/2017-03-22-protected-branches-admin-improvements/, they'll remove the deprecated `include_admins` field on May 2, 2017, so this Go API will become less confusing then.

So, this is okay by me, I won't request further improvements (I can think of some, but I don't think it's worth the overhead).

**Comment by shurcooL** [comment]
Thanks a lot @amirilovic!

Based on what it says at https://developer.github.com/changes/2017-03-22-protected-branches-admin-improvements/, we'll need to wait until May 2, 2017 to be able to completely resolve #606.
